### DISCUSSION TITLE
How to implement a telemetry gateway?

### DISCUSSION THREAD

**Discussion by zigzag312** [open-telemetry/opentelemetry-dotnet#5354] [Q&A] [answered]
I have ASP.NET Core endpoint that receives log events from a mobile app (in custom format). 

From there, I would like to forward these events to an OTEL endpoint.

Using `ILogger` I'm not able to create a correct resource with client's attributes (e.g. device mode, OS version etc.). 

I also tried to use gRPC's `LogsServiceClient` class directly, but I can't as `LoggerService` class is marked as internal.

Is there a way to accomplish this task using opentelemetry-dotnet or do I need to generate my own `LogsServiceClient` from .proto files?

**Comment by vishweshbankwar** [comment]
To send data to an otlp endpoint, you would need to use otlp log exporter. 

https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Exporter.OpenTelemetryProtocol

**Reply by zigzag312** [reply]
But how can I add a log event with Resource's attributes about the client (mobile phone)? If I use `ILogger`, Resource attributes will be of the server by default.

**Reply by vishweshbankwar** [reply]
Just to confirm you are referring to https://opentelemetry.io/docs/concepts/resources?

**Reply by zigzag312** [reply]
Yes. The entities producing telemetry are users' mobile phones. So, resource attributes are mobile app version, mobile phone model, Android/iOS version etc. 

The task of ASP.NET Core server is to receive these events, transform them to open telemetry format and forward them to Otlp endpoint.

**Reply by cijothomas** [reply, marked as answer]
You probably need to create own clients for the same (the proto files for OTLP are here, and can generate clients and use it to achieve the same.).

**Reply by zigzag312** [reply]
@cijothomas Thanks, I did that today. I wasn't sure about the appropriate approach to do this, but it seems this is the way.
